### ISSUE TITLE
[Error] expected Tensor as element 1 in argument 0, but got int

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs4anki#278]
I am trying to implement the BW metric for calculating difficulty in the optimizer.

However, I am getting the following error. Could you please help me resolve it?

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/5df1bae6-b5ea-4fd9-9138-8b257a6a4287)
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/9af6a673-4d09-4044-9b0f-4ea3f58e5f17)

Code:
```py
    def step(self, X, state):
        '''
        :param X: shape[batch_size, 2], X[:,0] is elapsed time, X[:,1] is rating
        :param state: shape[batch_size, 2], state[:,0] is stability, state[:,1] is difficulty
        :return state:
        '''
        if torch.equal(state, torch.zeros_like(state)):
            # first learn, init memory states
            new_s = self.w[0] + self.w[1] * (X[:,1] - 1)
            new_d = 5
        else:
            r = torch.exp(np.log(0.9) * X[:,0] / state[:,0])
            if X[:,1] > 1:
              bw = 1 - r
            else:
              bw = 0 - r

            if bw >= 0.1:
              rd = 0
            elif bw <= -0.9:
              rd = 1
            else:
              rd = 0.1 - bw

            new_d = 10(self.w[13] * rd + (1 - self.w[13]) * state[:,1] / 10)
            new_d = new_d.clamp(0, 10)
            
            condition = X[:,1] > 1
            new_s = torch.where(condition, self.stability_after_success(state, new_d, r), self.stability_after_failure(state, new_d, r))
        new_s = new_s.clamp(0.1, 36500)
        return torch.stack([new_s, new_d], dim=1)
```

**Comment by L-M-Sherlock** [comment]
> ```python
> 
>             new_d = 5
> ```

It should be a tensor with the same shape, not a int.

**Comment by user1823** [comment]
> > ```python
> >             new_d = 5
> > ```
> 
> It should be a tensor with the same shape, not a int.

Ok. I replaced with the original formula for initial difficulty.

But, now I am getting the following error:

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/60fb20cd-11de-41a0-abb9-a181d2a1795e)

The code I am currently using is:
```py
def step(self, X, state):
        '''
        :param X: shape[batch_size, 2], X[:,0] is elapsed time, X[:,1] is rating
        :param state: shape[batch_size, 2], state[:,0] is stability, state[:,1] is difficulty
        :return state:
        '''
        if torch.equal(state, torch.zeros_like(state)):
            # first learn, init memory states
            new_s = self.w[0] + self.w[1] * (X[:,1] - 1)
            new_d = self.w[2] - self.w[3] * (X[:,1] - 3)
            new_d = new_d.clamp(0, 10)
        else:
            r = torch.exp(np.log(0.9) * X[:,0] / state[:,0])
            condition = X[:,1] > 1

            bw = torch.where(condition, 1 - r, 0 - r)
            
            if bw >= 0.1:
              rd = 0
            elif bw <= -0.9:
              rd = 1
            else:
              rd = 0.1 - bw

            new_d = 10(self.w[13] * rd + (1 - self.w[13]) * state[:,1] / 10)
            new_d = new_d.clamp(0, 10)

            new_s = torch.where(condition, self.stability_after_success(state, new_d, r), self.stability_after_failure(state, new_d, r))
        new_s = new_s.clamp(0.1, 36500)
        return torch.stack([new_s, new_d], dim=1)
```

**Comment by L-M-Sherlock** [comment]
>             bw = torch.where(condition, 1 - r, 0 - r)
>             
>             if bw >= 0.1:
>               rd = 0
>             elif bw <= -0.9:
>               rd = 1
>             else:
>               rd = 0.1 - bw

`bw` is a tensor of bool instead of a value here. So `if bw` is ambiguous. You should use `torch.where` for the subsequent operation in `rd`, too.

**Comment by user1823** [comment]
I don't know how to do that. Could you help me re-write this?

**Comment by user1823** [comment]
Never mind. I managed to re-write the condition by taking some help from some websites.

But, now I am getting the following error. I hope that this is the last one.

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/2aa8b967-df86-4b1f-8fc7-b0723593617b)
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/2c9394c6-3646-46e5-a059-fa935a0bda92)

My current code is:
```py
    def step(self, X, state):
        '''
        :param X: shape[batch_size, 2], X[:,0] is elapsed time, X[:,1] is rating
        :param state: shape[batch_size, 2], state[:,0] is stability, state[:,1] is difficulty
        :return state:
        '''
        if torch.equal(state, torch.zeros_like(state)):
            # first learn, init memory states
            new_s = self.w[0] + self.w[1] * (X[:,1] - 1)
            new_d = self.w[2] - self.w[3] * (X[:,1] - 3)
            new_d = new_d.clamp(0, 10)
        else:
            r = torch.exp(np.log(0.9) * X[:,0] / state[:,0])
            condition = X[:,1] > 1

            bw = torch.where(condition, 1 - r, 0 - r)

            rd = torch.where(bw >= 0.1, 0, torch.where(bw <= -0.9, 1, 0.1 - bw))

            new_d = 10(self.w[13] * rd + (1 - self.w[13]) * state[:,1] / 10)
            new_d = new_d.clamp(0, 10)

            new_s = torch.where(condition, self.stability_after_success(state, new_d, r), self.stability_after_failure(state, new_d, r))
        new_s = new_s.clamp(0.1, 36500)
        return torch.stack([new_s, new_d], dim=1)
```

**Comment by L-M-Sherlock** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/ea052e6b-accd-424f-95b7-cedb4838880e)

**Comment by user1823** [comment]
> ![image](https://user-images.githubusercontent.com/32575846/243167864-ea052e6b-accd-424f-95b7-cedb4838880e.png)

Thanks for that. I really thought that the multiplication sign is unnecessary before a bracket (just like mathematics).

But, unfortunately, I am still getting an error.

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/e4321e29-972b-413c-8e07-c38886d666fe)

**Comment by L-M-Sherlock** [comment]
`new_d` should not be zero. Because this function:

```
    def stability_after_failure(self, state, new_d, r):
        new_s = self.w[9] * torch.pow(new_d, -self.w[10]) * torch.pow(
            state[:,0], self.w[11]) * torch.exp((1 - r) * self.w[12])
        return new_s
```

When `new_d` is zero, `torch.pow(new_d, -self.w[10])` will be `inf`. It could cause the error.
